File: graphs/1971_find_if_path_exists_in_graph.ts
function validPath(n: number, edges: number[][], source: number, destination: number): boolean {

    let seen = new Array(n).fill(false);
    let mapp = {};

    for(let [a,b] of edges){

        let currA = (mapp[a] || []);
        currA.push(b);
        mapp[a] = currA;
        let currB = (mapp[b] || []);
        currB.push(a);
        mapp[b] = currB;
    }

    let prev = [source];
    let curr = [];

    while(prev.length != 0){

        for(let p of prev){
            if(p === destination){
                return true;
            }
            if(seen[p] == true){
                continue;
            } else {
                seen[p] = true;
            }

            curr.push(... mapp[p]);
        }

        prev = curr;
        curr = [];
    }

    return false;

};

File: graphs/1971_find_if_path_exists_in_graph.rs
use std::collections::{HashMap, HashSet};

impl Solution {
    pub fn valid_path(n: i32, edges: Vec<Vec<i32>>, source: i32, destination: i32) -> bool {
        let mut seen: Vec<bool> = vec![false; n as usize];
        let mut mapp: HashMap<i32, Vec<i32>> = HashMap::new();

        for edge in edges {
            let a = edge[0];
            let b = edge[1];
            let curr_a = mapp.entry(a).or_insert(Vec::new());
            curr_a.push(b);
            let curr_b = mapp.entry(b).or_insert(Vec::new());
            curr_b.push(a);
        }

        let mut prev = HashSet::new();
        prev.insert(source);
        let mut curr = HashSet::new();

        while !prev.is_empty() {
            for &p in &prev {
                if p == destination {
                    return true;
                }
                if seen[p as usize] {
                    continue;
                }
                seen[p as usize] = true;

                if let Some(routes) = mapp.get(&p) {
                    for &neighbor in routes {
                        curr.insert(neighbor);
                    }
                }
            }

            prev = std::mem::take(&mut curr);
        }

        false
    }
}
